File: src/claw_mem/dreaming/deep.py
# Copyright 2026 Peter Cheng
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""
Dreaming Engine — Deep Phase (Candidate Scorer | v4.12.0)

Six-dimensional heuristic scoring with weighted composite.
Scores normalized to 0.0–1.0, weighted by DreamingConfig.
"""

from __future__ import annotations

import math
import re
from dataclasses import dataclass, field
from datetime import datetime, timezone
from typing import Any, Dict, List, Optional, TYPE_CHECKING

if TYPE_CHECKING:
    from .config import DreamingConfig
    from .light import Signal


@dataclass
class ScoredCandidate:
    """A signal with all six dimension scores and a weighted composite.

    Attributes:
        signal: The original staged signal.
        frequency_score: 0.0-1.0, based on recall_count.
        relevance_score: 0.0-1.0, from prior relevance_scores.
        query_diversity_score: 0.0-1.0, based on unique_queries.
        recency_score: 0.0-1.0, temporal freshness.
        consolidation_score: 0.0-1.0, integration with existing knowledge.
        conceptual_richness_score: 0.0-1.0, information density.
        composite: Weighted sum of all six scores.
    """

    signal: Any  # Signal
    frequency_score: float = 0.0
    relevance_score: float = 0.0
    query_diversity_score: float = 0.0
    recency_score: float = 0.0
    consolidation_score: float = 0.0
    conceptual_richness_score: float = 0.0
    composite: float = 0.0

    def to_dict(self) -> Dict[str, Any]:
        return {
            "signal": self.signal.to_dict(),
            "frequency_score": round(self.frequency_score, 3),
            "relevance_score": round(self.relevance_score, 3),
            "query_diversity_score": round(self.query_diversity_score, 3),
            "recency_score": round(self.recency_score, 3),
            "consolidation_score": round(self.consolidation_score, 3),
            "conceptual_richness_score": round(self.conceptual_richness_score, 3),
            "composite": round(self.composite, 3),
        }


class CandidateScorer:
    """Six-dimensional heuristic scorer for dreaming candidates.

    Each dimension is independently scored 0.0–1.0, then combined
    using configured weights to produce a composite score.
    """

    def __init__(self, config: Optional[DreamingConfig] = None):
        from .config import DreamingConfig

        self._config = config or DreamingConfig()

    def score_all(self, signals: List) -> List[ScoredCandidate]:
        """Score all staged signals.

        Args:
            signals: List of Signal objects from the light phase.

        Returns:
            List of ScoredCandidate sorted by composite descending.
        """
        candidates: List[ScoredCandidate] = []

        for sig in signals:
            freq = self._score_frequency(sig)
            rel = self._score_relevance(sig)
            div = self._score_query_diversity(sig)
            rec = self._score_recency(sig)
            con = self._score_consolidation(sig)
            rich = self._score_conceptual_richness(sig)

            composite = (
                freq * self._config.frequency_weight
                + rel * self._config.relevance_weight
                + div * self._config.query_diversity_weight
                + rec * self._config.recency_weight
                + con * self._config.consolidation_weight
                + rich * self._config.conceptual_richness_weight
            )

            candidates.append(
                ScoredCandidate(
                    signal=sig,
                    frequency_score=freq,
                    relevance_score=rel,
                    query_diversity_score=div,
                    recency_score=rec,
                    consolidation_score=con,
                    conceptual_richness_score=rich,
                    composite=composite,
                )
            )

        candidates.sort(key=lambda c: c.composite, reverse=True)
        return candidates

    def filter(self, candidates: List[ScoredCandidate]) -> List[ScoredCandidate]:
        """Filter candidates by score threshold and top-k.

        Args:
            candidates: Scored candidates (already sorted by composite).

        Returns:
            Filtered list of ScoredCandidate.
        """
        above_threshold = [
            c for c in candidates if c.composite >= self._config.score_threshold
        ]
        return above_threshold[: self._config.top_k_candidates]

    # ── dimension scorers ──────────────────────────────────────────

    @staticmethod
    def _score_frequency(sig) -> float:
        """Score based on recall_count. Saturates at ~10."""
        return min(sig.recall_count / 10.0, 1.0)

    @staticmethod
    def _score_relevance(sig) -> float:
        """Average of existing relevance scores, defaulting to 0.5."""
        scores = sig.relevance_scores
        if not scores:
            return 0.5
        return sum(scores) / len(scores)

    @staticmethod
    def _score_query_diversity(sig) -> float:
        """Score based on unique query count. Saturates at ~5."""
        return min(sig.unique_queries / 5.0, 1.0)

    @staticmethod
    def _score_recency(sig) -> float:
        """Exponential decay from timestamp. Newer = higher score."""
        if not sig.timestamp:
            return 0.5
        try:
            ts = datetime.fromisoformat(sig.timestamp)
            now = datetime.now(timezone.utc).replace(tzinfo=None)
            ts_naive = ts.replace(tzinfo=None)
            age_hours = (now - ts_naive).total_seconds() / 3600.0
            # Clamp negative ages (future timestamps) to zero age
            if age_hours < 0:
                age_hours = 0.0
            # Half-life of 24 hours
            return math.exp(-age_hours / 24.0)
        except (ValueError, TypeError):
            return 0.5

    @staticmethod
    def _score_consolidation(sig) -> float:
        """Proxy: higher tag count suggests better integration."""
        tag_count = len(sig.tags) if sig.tags else 0
        return min(tag_count / 5.0, 1.0)

    @staticmethod
    def _score_conceptual_richness(sig) -> float:
        """Proxy: content length and named-entity count as information density."""
        text = sig.content or ""
        # Length component
        length_score = min(len(text) / 200.0, 1.0)
        # Named entity proxy: uppercase words as simple heuristic
        entities = len(re.findall(r"\b[A-Z][a-z]+\b", text))
        entity_score = min(entities / 3.0, 1.0)
        return (length_score * 0.6 + entity_score * 0.4)

File: src/dreaming/deep.ts
// Copyright 2026 Peter Cheng
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

/**
 * Dreaming Engine -- Deep Phase (Candidate Scorer | v4.12.0)
 *
 * Six-dimensional heuristic scoring with weighted composite.
 * Scores normalized to 0.0-1.0, weighted by DreamingConfig.
 */

import { DEFAULT_DREAMING_CONFIG, type DreamingConfig } from "./config";
import type { Signal } from "./light";

// ── ScoredCandidate ──────────────────────────────────────────────────────

export interface ScoredCandidate {
  /** The original staged signal. */
  signal: Signal;
  /** 0.0-1.0, based on recall_count. */
  frequencyScore: number;
  /** 0.0-1.0, from prior relevance_scores. */
  relevanceScore: number;
  /** 0.0-1.0, based on unique_queries. */
  queryDiversityScore: number;
  /** 0.0-1.0, temporal freshness. */
  recencyScore: number;
  /** 0.0-1.0, integration with existing knowledge. */
  consolidationScore: number;
  /** 0.0-1.0, information density. */
  conceptualRichnessScore: number;
  /** Weighted sum of all six scores. */
  composite: number;
}

export function createScoredCandidate(
  signal: Signal,
  overrides?: Partial<Omit<ScoredCandidate, "signal">>,
): ScoredCandidate {
  return {
    signal,
    frequencyScore: 0,
    relevanceScore: 0,
    queryDiversityScore: 0,
    recencyScore: 0,
    consolidationScore: 0,
    conceptualRichnessScore: 0,
    composite: 0,
    ...overrides,
  };
}

export function scoredCandidateToDict(c: ScoredCandidate): Record<string, unknown> {
  return {
    signal: {
      memory_id: c.signal.memoryId,
      content: c.signal.content,
      memory_type: c.signal.memoryType,
      recall_count: c.signal.recallCount,
      unique_queries: c.signal.uniqueQueries,
      relevance_scores: c.signal.relevanceScores,
      tags: c.signal.tags,
      timestamp: c.signal.timestamp,
    },
    frequency_score: round3(c.frequencyScore),
    relevance_score: round3(c.relevanceScore),
    query_diversity_score: round3(c.queryDiversityScore),
    recency_score: round3(c.recencyScore),
    consolidation_score: round3(c.consolidationScore),
    conceptual_richness_score: round3(c.conceptualRichnessScore),
    composite: round3(c.composite),
  };
}

function round3(v: number): number {
  return Math.round(v * 1000) / 1000;
}

// ── CandidateScorer ──────────────────────────────────────────────────────

export class CandidateScorer {
  private _config: DreamingConfig;

  constructor(config?: DreamingConfig) {
    this._config = config ?? DEFAULT_DREAMING_CONFIG;
  }

  /**
   * Score all staged signals.
   *
   * @param signals - List of Signal objects from the light phase.
   * @returns List of ScoredCandidate sorted by composite descending.
   */
  scoreAll(signals: Signal[]): ScoredCandidate[] {
    const candidates: ScoredCandidate[] = [];

    for (const sig of signals) {
      const freq = this._scoreFrequency(sig);
      const rel = this._scoreRelevance(sig);
      const div = this._scoreQueryDiversity(sig);
      const rec = this._scoreRecency(sig);
      const con = this._scoreConsolidation(sig);
      const rich = this._scoreConceptualRichness(sig);

      const composite =
        freq * this._config.frequencyWeight +
        rel * this._config.relevanceWeight +
        div * this._config.queryDiversityWeight +
        rec * this._config.recencyWeight +
        con * this._config.consolidationWeight +
        rich * this._config.conceptualRichnessWeight;

      candidates.push({
        signal: sig,
        frequencyScore: freq,
        relevanceScore: rel,
        queryDiversityScore: div,
        recencyScore: rec,
        consolidationScore: con,
        conceptualRichnessScore: rich,
        composite,
      });
    }

    candidates.sort((a, b) => b.composite - a.composite);
    return candidates;
  }

  /**
   * Filter candidates by score threshold and top-k.
   *
   * @param candidates - Scored candidates (already sorted by composite).
   * @returns Filtered list of ScoredCandidate.
   */
  filter(candidates: ScoredCandidate[]): ScoredCandidate[] {
    const aboveThreshold = candidates.filter(
      (c) => c.composite >= this._config.scoreThreshold,
    );
    return aboveThreshold.slice(0, this._config.topKCandidates);
  }

  // ── dimension scorers ──────────────────────────────────────────────────

  /** Score based on recall_count. Saturates at ~10. */
  private _scoreFrequency(sig: Signal): number {
    return Math.min(sig.recallCount / 10.0, 1.0);
  }

  /** Average of existing relevance scores, defaulting to 0.5. */
  private _scoreRelevance(sig: Signal): number {
    const scores = sig.relevanceScores;
    if (!scores || scores.length === 0) return 0.5;
    return scores.reduce((a, b) => a + b, 0) / scores.length;
  }

  /** Score based on unique query count. Saturates at ~5. */
  private _scoreQueryDiversity(sig: Signal): number {
    return Math.min(sig.uniqueQueries / 5.0, 1.0);
  }

  /** Exponential decay from timestamp. Newer = higher score. */
  private _scoreRecency(sig: Signal): number {
    if (!sig.timestamp) return 0.5;
    try {
      const ts = new Date(sig.timestamp);
      const now = new Date();
      const ageMs = now.getTime() - ts.getTime();
      const ageHours = Math.max(0, ageMs / 3600000);
      // Half-life of 24 hours
      return Math.exp(-ageHours / 24.0);
    } catch {
      return 0.5;
    }
  }

  /** Proxy: higher tag count suggests better integration. */
  private _scoreConsolidation(sig: Signal): number {
    const tagCount = sig.tags?.length ?? 0;
    return Math.min(tagCount / 5.0, 1.0);
  }

  /**
   * Proxy: content length and capitalized-word count as information density.
   */
  private _scoreConceptualRichness(sig: Signal): number {
    const text = sig.content ?? "";
    // Length component
    const lengthScore = Math.min(text.length / 200.0, 1.0);
    // Named entity proxy: uppercase words as simple heuristic
    const entities = (text.match(/\b[A-Z][a-z]+\b/g) ?? []).length;
    const entityScore = Math.min(entities / 3.0, 1.0);
    return lengthScore * 0.6 + entityScore * 0.4;
  }
}
